Enhance sun component

pnbruckner opened the pull request:
## Description:
This pull request proposes some useful enhancements to the sun component. Without any configuration changes the new component will function exactly as the current component, so it is fully backward compatible, and hence is not a "breaking change."

The first change is to add some new, optional attributes - specifically daylight, next_daylight, prev_daylight, sunrise and sunset.

The sunrise and sunset attributes are particularly useful for paralleling the existing sun conditions in template code. Unfortunately the existing next_rising and next_setting attributes do not work well in this scenario because they change as soon as sunrise and sunset are passed, respectively. The new attributes stay constant throughout the day (only updating at midnight local), so they can be used effectively in templates.

The next change allows optional selection of which attributes the sun component maintains via a new monitored_conditions configuration parameter. The default is to maintain the original set of attributes. Since the component must update at each maintained attribute/event, if only a subset of the attributes are of interest to the user, there is really no need to take time to update the component for a data point the user does not use. This can be particularly useful if the user does not use or need azimuth and elevation, since these attributes are updated every minute.

Which leads to the next change, the addition of an optional scan_interval configuration parameter. If azimuth and/or elevation are being used, then this parameter controls how often they are updated. The default is to implement the existing behavior. If specified the rate at which these attributes are updated can be reduced, resulting in a potentially significant reduction in corresponding state_changed events. Of course, if azimuth and elevation are not used or needed, then all of the corresponding, frequent updates can be eliminated entirely.

To these ends the following, specific changes were implemented:

Add monitored_conditions config item to specify which attributes to maintain. (next_rising and next_setting will always be maintained.) If monitored_conditions is not specified then original set of attributes will be maintained. For elevation and azimuth, add scan_interval (minimum of one minute.) Add optional new attributes daylight, next_daylight, prev_daylight, sunrise & sunset, all of which will be updated at midnight local.

Additional background and discussion can be found [here](https://community.home-assistant.io/t/hours-of-daylight/59422) and [here](https://community.home-assistant.io/t/sensor-sunrise-and-sensor-sunset-in-sync-for-todays-date/63191/2).

**Related issue (if applicable):** none

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#6053

## Example entry for `configuration.yaml` (if applicable):
```yaml
sun:
  monitored_conditions:
    - elevation
    - sunrise
    - sunset
  scan_interval:
    minutes: 10
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

pnbruckner committed:
Update sun.py

Add monitored_conditions config item to specify which attributes to maintain. If not specified then original set will be used. For elevation and azimuth, add scan_interval (minimum of one minute.) If next_rising or next_setting are not specified they will not be added as attributes. However, they will always be maintained so that sun's state can be determined/maintained. Add optional new attributes daylight, next_daylight, prev_daylight, sunrise & sunset, all of which will be updated at midnight local.

homeassistant commented:
Hi @pnbruckner, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#15961).

Once you do that we will be able to review and accept this pull request.

Thanks!

pnbruckner committed:
Fix Travis CI build failures

pnbruckner commented:
I tried to fix the under and over indention failures but I don't seem to understand what it wants. If someone could point me to the coding standard where this is defined, I'll try again.

pnbruckner committed:
Fix some lint and pylint failures

pnbruckner committed:
Fix more test failures

Allow extra config items for outdated elevation

balloob commented:
The test failure is because no tests were added for the new functionality.

balloob commented:
I'm glad someone is taking the time to kill the state change events every minute, such a spamming feature.

pnbruckner commented:
I'm still pretty new to HA. This is my first code pull request. I haven't yet [setup the development environment](https://developers.home-assistant.io/docs/en/development_environment.html), and I haven't even begun to look at how to add test cases. I'll start on that tomorrow. I have over 40 years of software development experience, but not a whole lot with HA. Any pointers, etc. are appreciated. Thanks!

DavidFW1960 commented:
Great work by @pnbruckner > Been using this as a custom component and it's a great enhancement.

awarecan commented on homeassistant/components/sun.py:
```diff
@@ -55,16 +83,33 @@ class Sun(Entity):
+            setattr(self, attr, None)
+        # Make sure STATE_ATTR_NEXT_RISING & STATE_ATTR_NEXT_SETTING are
+        # defined since they are used to calculate state. Initialize them.
+        setattr(self, STATE_ATTR_NEXT_RISING, None)
```
Since these two attributes are must have, can we just
```python
self.next_rising = None
```

pnbruckner replied:
Yes, that's another way to do it. But then I'd also have to change the state_attributes(), next_change() and update_as_of() methods as well. According to the Python docs the two ways are the same: "For example, setattr(x, 'foobar', 123) is equivalent to x.foobar = 123." So I don't see an advantage to changing this. I think the way it is is more consistent. But if you feel strongly about it, I can certainly change it.

awarecan replied:
No problem, it is just a comment.

pnbruckner commented:
I set up the development environment on a clean Linux Mint virtual machine and switched to my patch-1 branch. FYI, I discovered the docs didn't mention the following components that I had to install to get tox to work: autoconf g++. And I had to install git. After doing all of this, I finally realized that it wasn't the tox testing that failed due to insufficient test cases - it was coverage/coveralls. Hmm, the docs don't mention how to install and use that. I guess I'll try to add some tests to tests/components/test_sun.py and try again.

BTW, actually the tox typing tests failed with a fairly useless error message:

ERROR: InvocationError for command '/bin/bash -c mypy homeassistant/*.py homeassistant/util/' (exited with code 1)

awarecan commented:
If you look at the coveralls result, "NEW MISSED LINES IN DIFF" shows up the codes missed in unit test.

If you still failed quality gate in coveralls, you can try to rebase your branch against dev.

pnbruckner committed:
Add tests for new features

pnbruckner commented:
coveralls details show the component I'm working on - sun.py - now has 100% coverage. So I'm going to start working on updating the documentation.

pnbruckner commented:
Hi all. I realize this is just one of 86 open pull requests, and there are only so many "members" to go around. But since this is my first code pull request, I just wanted to make sure there wasn't anything else I need to do. E.g., should I remove the lines under "If the code communicates with devices, web services, or third-party tools:" in the description, since these don't apply? Or should I simply "check" them? As it is, I see this results in "4 of 8" showing under the title of the PR, which might lead people to think I'm not done yet. Thanks!

awarecan commented:
Code looks good to me. @MartinHjelmare ?

Just one suggestion, if `next_rising`, `next_setting` and `elevation` has to be monitored for front end and for backward compatibility, shall we remove them from `monitored_conditions` list 

By the way, 3 days is not too long for new pull request be reviewed.

pnbruckner commented:
Thanks @awarecan. Yep, not concerned about how long the process takes. Just wanted to make sure I wasn't the hold up.

If elevation was always maintained, then one of the biggest advantages of these changes would be lost - i.e., reducing the large number of state_changed events the sun component causes. And if a user doesn't care about elevation, or need it for something else, what does it hurt for it to be blank in the UI? For those who do want to see it or need it for something else, they can simply add it to the monitored_conditions list.

BTW, unless someone specifically _adds_ monitored_conditions to their sun configuration, all the previous attributes, including next_rising, next_setting & elevation will be maintained anyway. So a user would have to explicitly add monitored_conditions before they could possibly lose these attributes. I did that purposely to maintain backward compatibility.

As for next_rising & next_setting, I could see removing them from the monitored_conditions and always having them as attributes. If that's the consensus I'm happy to do that.

pvizeli commented on homeassistant/components/sun.py:
```diff
@@ -7,14 +7,18 @@
 import asyncio
 import logging
 from datetime import timedelta
+import voluptuous as vol
```
please follow PEP8 with ordering

pnbruckner replied:
Yep, sorry, missed that.

pvizeli commented on homeassistant/components/sun.py:
```diff
@@ -55,16 +83,33 @@ class Sun(Entity):
+        # Make sure STATE_ATTR_NEXT_RISING & STATE_ATTR_NEXT_SETTING are
+        # defined since they are used to calculate state. Initialize them.
+        setattr(self, STATE_ATTR_NEXT_RISING, None)
+        setattr(self, STATE_ATTR_NEXT_SETTING, None)
```
Please use our code style: `self.xy = bla`
setattr is used by dynamic property allocation.

pvizeli commented on homeassistant/components/sun.py:
```diff
@@ -55,16 +83,33 @@ class Sun(Entity):
+                # If scan_interval not specified, use old method of updating
+                # once a minute on the half minute (i.e., now == xx:xx:30.)
+                async_track_utc_time_change(hass, self.timer_update,
+                                            second=30)
```
Please migrate all things they need `hass` into `async_added_to_hass` function. That is the new style.

pnbruckner replied:
How can that work here? Isn't calling async_added_to_hass a feature of entity_platform.py? But the sun component doesn't use entity_platform.py. I'm not sure how to do what you suggest.

pvizeli commented on homeassistant/components/sun.py:
```diff
@@ -55,16 +83,33 @@ class Sun(Entity):
-        async_track_utc_time_change(hass, self.timer_update, second=30)
+        # Initialize values for attributes to be reported.
+        for attr in self._monitored_condtions:
+            setattr(self, attr, None)
```
Please don't do that. Use a internal dict as cache, if you need one. It make all things hard to read and hard to read mean that other make mistakes. Keep it simple

pnbruckner replied:
Good idea. Thanks! Will do.

pvizeli commented on homeassistant/components/sun.py:
```diff
@@ -74,57 +119,108 @@ def name(self):
+                     STATE_ATTR_NEXT_DAYLIGHT]:
+            if attr in self._monitored_condtions:
+                attrs[attr] = getattr(self, attr)
+        return attrs
```
Keep it simple and remove `getattr` handling

pnbruckner replied:
Yep, this will change per the other suggestions.

pvizeli reviewed (changes requested):
First round

@awarecan if there is a PR taged with `core`, it need a core review too

pnbruckner committed:
Update per review change requests

Format imports per PEP8. Revert next_rising and next_setting to
standard attributes and remove from monitored_conditions options. Use
new dictionary to manage optional attributes. Simplify methods
state_attributes and update_as_of to make cleaner and easier to
understand/maintain. In method update_sun_position, only update
monitored attributes.

pnbruckner commented:
I updated the code per the review change requests. It might be easier to compare to the original version, since I reverted some code back to the way it was. Also some of my new code should be simpler and easier to understand, especially as compared to the original code.

pnbruckner commented:
@pvizeli, just wondering if my last update satisfies your change requests. Also, what do you want me to do about async_added_to_hass? Per my reply I can't see how I can do what you suggested regarding that function. Thanks! If this is still on your to-do list, that's fine. Again I'm new to this so I just wanted to make sure this wasn't still waiting on something from me.

quthla commented:
@pvizeli Can this be merged now?

pnbruckner commented:
@balloob, sorry, I know you're a busy guy, but is it common to wait this long? I first submitted this over two weeks ago. Then four days later (which, of course, was not an unreasonable wait) changes were requested. I provided changes the same day. But now I've been waiting a week and a half with no update, and not even a courtesy reply to my questions.

It's not necessarily the amount of time. It's getting no response at all that is disheartening. I've got other ideas of where I can contribute, but if this is any indication of how things go, I'm not sure it's worth the effort. I understand that this is a spare time project for most people, and I'm happy to wait my turn, but waiting with no end in sight is not encouraging.

Thanks for listening!

balloob commented:
I've thought about it. Allowing picking of the attributes that the sun entity will contain is against how the rest of the system works. Monitored conditions should control the different individual entities that are being published. The only reason this one is called `sun.sun` is because it's one of the very first components ever built for Home Assistant. Before there even was a concept of (binary) sensors.

I think that this functionality should be added as new platforms for sensor and binary sensor (above horizon/below horizon). We should probably leave the old component alone just to not do any breaking changes. Things like daylight, next setting etc can all be published under a new device class [timestamp](https://github.com/home-assistant/architecture/issues/39).

pnbruckner commented:
Ok, fair enough. But I'm not up to creating a whole new set of platforms, so I'll just close this PR. Feel free to use any of my ideas or code. Thanks.
